cosmos.go

```diff
@@ -203,39 +217,133 @@ func (c *cosmosImpl) dial() (interfaces.QueryExecutor, error) {
 
 func (c *cosmosImpl) ExecuteQuery(query interfaces.QueryBuilder) ([]interfaces.Response, error) {
 	if query == nil {
-		return nil, fmt.Errorf("Query is nil")
+		return nil, fmt.Errorf("query is nil")
 	}
 	return c.Execute(query.String())
 }
 
 func (c *cosmosImpl) Execute(query string) ([]interfaces.Response, error) {
 
-	responses, err := c.pool.Execute(query)
+	doRetry := func() ([]interfaces.Response, error) {
+		return c.pool.Execute(query)
+	}
+
+	responses, err := c.retryLoop(doRetry)
 
 	// try to investigate the responses and to find out if we can find more specific error information
 	if respErr := extractFirstError(responses); respErr != nil {
 		err = respErr
 	}
 
-	updateRequestMetrics(responses, c.metrics)
 	return responses, err
 }
 
 func (c *cosmosImpl) ExecuteWithBindings(query string, bindings, rebindings map[string]interface{}) ([]interfaces.Response, error) {
 
-	responses, err := c.pool.ExecuteWithBindings(query, bindings, rebindings)
+	doRetry := func() ([]interfaces.Response, error) {
+		return c.pool.ExecuteWithBindings(query, bindings, rebindings)
+	}
+
+	responses, err := c.retryLoop(doRetry)
 
 	// try to investigate the responses and to find out if we can find more specific error information
 	if respErr := extractFirstError(responses); respErr != nil {
 		err = respErr
 	}
 
-	updateRequestMetrics(responses, c.metrics)
+	return responses, err
+}
+
+type retryFun func() ([]interfaces.Response, error)
+
+func (c *cosmosImpl) retryLoop(executeRetry retryFun) (responses []interfaces.Response, err error) {
+	var retryCount int
+
+	for retryCount = 0; retryCount < c.maxRetries+1; retryCount++ {
+
+		responses, err = executeRetry()
+
+		updateRequestMetrics(responses, c.metrics)
+
+		if !c.automaticRetries || err != nil {
+			break
+		}
+
+		retryInformation := extractRetryConditions(responses)
+
+		// Retry is always on a new or at least active connection,
+		// therefore retryInformation.retryOnNewConnection can be used here as well
+		if !(retryInformation.retry || retryInformation.retryOnNewConnection) {
+			break
+		}
+
+		c.logger.Debug().Msgf("retrying query after %v because of header status code %d", retryInformation.retryAfter, retryInformation.responseStatusCode)
+		if retryInformation.retryAfter > 0 {
+			time.Sleep(retryInformation.retryAfter)
```

Comment:
this is dangerous if `retryAfter` is big (and it comes from an external service) then the whole call and even the graceful shutdown of the application is blocked for x seconds.

We should at least implement a timeout here to wait at max for X seconds (configurable via `gremcos.AutomaticRetries()` option.